### PR TITLE
MAINT: Move ``NPY_VSTRING`` and make ``NPY_NTYPES NPY_TYPES_LEGACY`` 

### PR DESCRIPTION
**Pull request by seberg**
This requires type numbers a bit, the main reason is that I want to be able to recognize "old style" dtypes based on something on the instance, and the type number is the obvious choice.

@ngoldbaum I would appreciate if we can put this in quickly and tweak if needed to help me push the big refactor.

### CODE REVIEW

File: doc/release/upcoming_changes/25347.c_api.rst [hunk 1]
```diff
@@ -2,8 +2,5 @@
   exposed. This includes functions for acquiring and releasing mutexes locking
   access to the string data as well as packing and unpacking UTF-8 bytestreams
   from array entries.
-* ``NPY_NTYPES`` is now version-dependent to accomadate the availability of a
-  new NumPy built-in DType. It is now defined in ``npy2_compat.h``.
-* ``NPY_NTYPES_LEGACY`` is now defined in ``ndarraytypes.h``, use this value
-  if you do not wish to update code to support dtypes written using the new
-  DType API that may not function in the same way as legacy dtypes.
+* ``NPY_NTYPES`` has been renamed to ``NPY_NTYPES_LEGACY`` as it does not
+  include new NumPy built-in DTypes (specially the string DType).
```
Comment (ngoldbaum) [inline doc/release/upcoming_changes/25347.c_api.rst:6]:
```suggestion
  include new NumPy built-in DTypes. In particular the new string DType
  will likely not work correctly with code that handles legacy DTypes.
```

Would it maybe be useful for downstream to have a `NTYPES_NONLEGACY` that is set to 1 now? I guess we can always add that later.

**Review by ngoldbaum — APPROVED** [review]
I struggled with whether or not to do it this way when I added `VSTRING` in the first place. I think it's totally fine to get rid of `NPY_NTYPES` and try to encourage downstream to handle new dtypes with more forward looking APIs by explicitly not including it in `NPY_NTYPES`. If doing it this way makes it easier to modify the descriptor struct, please do so.

Just a few comments and questions below.

File: doc/release/upcoming_changes/25347.c_api.rst (hunk 1 above)
Comment (seberg) [reply doc/release/upcoming_changes/25347.c_api.rst:6]: Yah, let's add it later, right now I am not even sure that `NPY_NTYPES` has a single user.

File: numpy/_core/src/multiarray/arraytypes.c.src [hunk 2]
```diff
@@ -4232,8 +4232,10 @@ NPY_NO_EXPORT PyArray_Descr @from@_Descr = {
 
 /**end repeat**/
 
-#define _MAX_LETTER 128
-static char _letter_to_num[_MAX_LETTER];
+/* The smallest type number is ?, the largest bounded by 'z'. */
+#define _MAX_LETTER ('z' + 1)
+static npy_int16 _letter_to_num[_MAX_LETTER - '?'];
```
[the remark this answers is not in the document]
Comment (seberg) [reply numpy/_core/src/multiarray/arraytypes.c.src:4237]: Yeah, it's super ugly and should be reorganized or just be a giant switch statement here, but not today.

File: numpy/_core/include/numpy/npy_2_compat.h
[the remark this answers is not in the document]
Comment (seberg) [reply numpy/_core/include/numpy/npy_2_compat.h:1]: Good idea, force pushed this on top of the other suggestions.

**Comment by seberg** [comment]
OK, had added that fix and i tis a tiny fixup, so putting this in.

**Pull request merged**